### DISCUSSION TITLE
Understanding the secret_provision_start_server API

### DISCUSSION THREAD

**Discussion by ishaangandhi** [gramineproject/gramine#1294] [Q&A] [answered]
Are measurements from the SGX quotes verified _before_ the secret is sent with the `secret_provision_start_server API`? That's what would make sense to me logically. Additionally, based on my understanding of RA-TLS, the connection only opens once the measurements are verified (per [Gramine's docs](https://gramine.readthedocs.io/en/stable/attestation.html#:~:text=Similarly%20to%20ra_tls_verify_epid,not%20thread%2Dsafe.)).

However, I see in the source code that you are supposed to check against "expected values."

https://github.com/gramineproject/gramine/blob/af7caa9813029c2b5a4656de098113a18863eda2/CI-Examples/ra-tls-secret-prov/secret_prov/server.c#L46

If that is true, how would you even check against expected values? Doesn't this defeat the point of the security policy?

Thanks!

**Comment by ishaangandhi** [comment]
Additionally, the docs for the `verify_measurements_cb_t m_cb` say:

```
Callback for user-specific verification of measurements in client's SGX
quote. If user supplies NULL, then default logic of RA-TLS is invoked.
```

Does this mean you can use the default logic of RA-TLS _or_ provide your own logic? Is there a way to use the default logic and _then_ your logic to, for example, verify the identity of the enclave with `MRENCLAVE`? That would seem very useful to me.

**Comment by ishaangandhi** [comment, marked as answer]
Oh I get it now. If you pass `NULL`, the required MRENCLAVE is taken from an environment variable, as I found out from the logs:

```
WARNING: The default enclave verification hook is being used, but RA_TLS_MRSIGNER is not set. This is deprecated and will become an error in the future. If you wish to accept any value, please specify RA_TLS_MRSIGNER=any explicitly.
WARNING: The default enclave verification hook is being used, but RA_TLS_MRENCLAVE is not set. This is deprecated and will become an error in the future. If you wish to accept any value, please specify RA_TLS_MRENCLAVE=any explicitly.
WARNING: The default enclave verification hook is being used, but RA_TLS_ISV_PROD_ID is not set. This is deprecated and will become an error in the future. If you wish to accept any value, please specify RA_TLS_ISV_PROD_ID=any explicitly.
WARNING: The default enclave verification hook is being used, but RA_TLS_ISV_SVN is not set. This is deprecated and will become an error in the future. If you wish to accept any value, please specify RA_TLS_ISV_SVN=any explicitly.
WARNING: Neither RA_TLS_MRSIGNER nor RA_TLS_MRENCLAVE are specified. This will accept any enclave and provides no security whatsoever.
```

**Comment by dimakuv** [comment]
One important thing to note here: `gramine/CI-Examples/ra-tls-secret-prov/secret_prov/server.c` is just an example. You are supposed to write your own secret provisioning server application for your own production usages. So you can replace the logic in this callback with your own logic: https://github.com/gramineproject/gramine/blob/af7caa9813029c2b5a4656de098113a18863eda2/CI-Examples/ra-tls-secret-prov/secret_prov/server.c#L36-L50

Ideally, you would replace the above dummy snippet with your own code that will check MRENCLAVE and other measurements against whatever you have as a policy and "golden" values.
